src/overlay/Peer.cpp

```diff
@@ -1868,7 +2024,199 @@ Peer::PeerMetrics::PeerMetrics(VirtualClock::time_point connectedTime)
+    mAdvertTimer.async_wait([this](asio::error_code const& error) {
+        if (!error)
+        {
+            flushAdvert();
+            startAdvertTimer();
+        }
+    });
+}
+
+void
+Peer::sendTxDemand(xdr::xvector<Hash>&& demands)
+{
+    ZoneScoped;
+    if (demands.size() > 0)
+    {
+        auto msg = std::make_shared<StellarMessage>();
+        msg->type(FLOOD_DEMAND);
+        msg->floodDemand().txHashes = std::move(demands);
+        std::weak_ptr<Peer> weak(static_pointer_cast<Peer>(shared_from_this()));
+        CLOG_TRACE(Overlay, "Peer::sendTxDemand -- demanding {} txns from {}",
+                   demands.size(), mApp.getConfig().toShortString(getPeerID()));
+        mApp.postOnMainThread(
+            [weak, msg = std::move(msg)]() {
+                auto strong = weak.lock();
+                if (strong)
+                {
+                    strong->sendMessage(msg);
+                }
+            },
+            "sendTxDemand");
+        getOverlayMetrics().mMessagesDemanded.Mark(demands.size());
+        TracyPlot("overlay.demand.size", static_cast<int64_t>(demands.size()));
+        ++mPeerMetrics.mTxDemandSent;
+    }
+}
+
+void
+Peer::flushAdvert()
+{
+    ZoneScoped;
+    auto& txHashes = mPendingAdvertMsg.floodAdvert().txHashes;
+
+    if (txHashes.size() > 0)
+    {
+        std::weak_ptr<Peer> weak(static_pointer_cast<Peer>(shared_from_this()));
+        auto msg = std::make_shared<StellarMessage>(mPendingAdvertMsg);
+        CLOG_TRACE(Overlay, "Peer::flushAdvert -- flushing {} tx hashes to {}",
+                   txHashes.size(),
+                   mApp.getConfig().toShortString(getPeerID()));
+        mApp.postOnMainThread(
+            [weak, msg = std::move(msg)]() {
+                auto strong = weak.lock();
+                if (strong)
+                {
+                    strong->sendMessage(msg);
+                }
+            },
+            "flushAdvert");
+        TracyPlot("overlay.advert.size", static_cast<int64_t>(txHashes.size()));
+
+        // Reset the advert.
+        mPendingAdvertMsg = StellarMessage();
+        mPendingAdvertMsg.type(FLOOD_ADVERT);
+    }
+}
+
+size_t
+Peer::txHashCount() const
+{
+    ZoneScoped;
+    return mIncomingTxHashes.size() + mTxHashesToRetry.size();
+}
+
+void
+Peer::appendTxHashesToRetry(std::list<Hash>& list)
+{
+    mTxHashesToRetry.splice(mTxHashesToRetry.end(), list);
+}
+
+Hash
+Peer::popFrontTxHash()
+{
+    ZoneScoped;
+    releaseAssert(txHashCount() > 0);
+
+    if (mTxHashesToRetry.size() > 0)
+    {
+        auto const h = mTxHashesToRetry.front();
+        mTxHashesToRetry.pop_front();
+        return h;
+    }
+    else
+    {
+        auto const h = mIncomingTxHashes.front();
+        mIncomingTxHashes.pop_front();
+        return h;
+    }
+}
+
+void
+Peer::fulfillDemand(FloodDemand const& dmd)
+{
+    ZoneScoped;
+    auto& herder = mApp.getHerder();
+    for (Hash h : dmd.txHashes)
```

Comment:
similar to adverts, should we limit to something like `FLOOD_MAX_DEMAND_SIZE` demands?

Reply:
(I know we limit downstream, but we should avoid doing work that we drop ourselves)

Reply:
I changed it to process up to `getLastMaxTxSetSizeOps` as that's the threshold flow control uses. I felt that `FLOOD_MAX_DEMAND_SIZE` is a bit aggressive for the same reason.

Reply:
This is also incorrect: you need to let flow control do the trimming with its own policy instead of trying to replicate it here. Right now the order of queuing is wrong because the trimming is premature, and you're relying on the limit defined in this function, which is error-prone.

Reply:
Removed.